debezium-connector-mongodb/src/main/java/io/debezium/connector/mongodb/MongoDbIncrementalSnapshotContext.java

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.connector.mongodb;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.Arrays;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.DebeziumException;
+import io.debezium.annotation.NotThreadSafe;
+import io.debezium.pipeline.source.snapshot.incremental.IncrementalSnapshotContext;
+import io.debezium.relational.Table;
+import io.debezium.util.HexConverter;
+
+/**
+ * A class describing current state of incremental snapshot of MongoDB connector
+ *
+ * @author Jiri Pechanec
+ *
+ */
+@NotThreadSafe
+public class MongoDbIncrementalSnapshotContext<T> implements IncrementalSnapshotContext<T> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbIncrementalSnapshotContext.class);
+
+    // TODO Consider which (if any) information should be exposed in source info
+    public static final String INCREMENTAL_SNAPSHOT_KEY = "incremental_snapshot";
+    public static final String DATA_COLLECTIONS_TO_SNAPSHOT_KEY = INCREMENTAL_SNAPSHOT_KEY + "_collections";
+    public static final String EVENT_PRIMARY_KEY = INCREMENTAL_SNAPSHOT_KEY + "_primary_key";
+    public static final String TABLE_MAXIMUM_KEY = INCREMENTAL_SNAPSHOT_KEY + "_maximum_key";
+
+    /**
+     * @code(true) if window is opened and deduplication should be executed
+     */
+    protected boolean windowOpened = false;
+
+    /**
+     * The last primary key in chunk that is now in process.
+     */
+    private Object[] chunkEndPosition;
+
+    // TODO After extracting add into source info optional block
+    // incrementalSnapshotWindow{String from, String to}
+    // State to be stored and recovered from offsets
+    private final Queue<T> dataCollectionsToSnapshot = new LinkedList<>();
+
+    private final boolean useCatalogBeforeSchema;
+    /**
+     * The PK of the last record that was passed to Kafka Connect. In case of
+     * connector restart the start of the first chunk will be populated from it.
+     */
+    private Object[] lastEventKeySent;
+
+    private String currentChunkId;
+
+    /**
+     * The largest PK in the table at the start of snapshot.
+     */
+    private Object[] maximumKey;
+
+    private Table schema;
```

Comment:
Surprising to see `Table` here (relational). I guess that's one thing you mean to address as part of that follow-up refactoring you mentioned?

Reply:
Yes, exactly, will be removed as part of incremental snapshot context refactoring.